subsys/bootloader/include/fw_metadata.h

```diff
@@ -59,24 +72,46 @@ struct __packed fw_validation_info {
 	u8_t  signature[CONFIG_SB_SIGNATURE_LEN];
 };
 
+/* Static asserts to ensure compatibility */
+static_assert(offsetof(struct fw_validation_info, magic) == 0,
+			"magic has wrong offset");
+static_assert(offsetof(struct fw_validation_info, firmware_address)
+			== CONFIG_SB_MAGIC_LEN,
+			"firmware_address has wrong offset");
+static_assert(offsetof(struct fw_validation_info, firmware_hash)
+			== (CONFIG_SB_MAGIC_LEN + 4),
+			"firmware_hash has wrong offset");
+static_assert(offsetof(struct fw_validation_info, public_key)
+			== (CONFIG_SB_MAGIC_LEN + 4 + CONFIG_SB_HASH_LEN),
+			"public_key has wrong offset");
+static_assert(offsetof(struct fw_validation_info, signature)
+			== (CONFIG_SB_MAGIC_LEN + 4 + CONFIG_SB_HASH_LEN
+			    + CONFIG_SB_SIGNATURE_LEN),
+			"signature has wrong offset");
+
 /*
  * Can be used to make the firmware discoverable in other locations, e.g. when
  * searching backwards. This struct would typically be constructed locally, so
  * it needs no version.
  */
-struct fw_validation_pointer {
+struct __packed fw_validation_pointer {
 	u32_t magic[MAGIC_LEN_WORDS];
 	const struct fw_validation_info *validation_info;
 };
 
+/* Static asserts to ensure compatibility */
+static_assert(offsetof(struct fw_validation_pointer, magic) == 0,
+			"magic has wrong offset");
+static_assert(offsetof(struct fw_validation_pointer, validation_info)
+			== CONFIG_SB_MAGIC_LEN,
+			"validation_info has wrong offset");
 
-static inline bool memeq_32(const void *expected, const void *actual,
-		u32_t len)
-{
-	u32_t *expected_32 = (u32_t *) expected;
-	u32_t *actual_32   = (u32_t *) actual;
 
-	__ASSERT(!(len & 3), "Length is not multiple of 4");
+/* All parameters must be word-aligned */
+static inline bool memeq_32(const void *expected, const void *actual, u32_t len)
+{
+	const u32_t *expected_32 = (const u32_t *) expected;
```

Comment:
`__ASSERT` discarded on purpose?

Reply:
Sort of, since it has moved to the `if` in `memeq()`. But I guess there's no reason not to have an `__ASSERT` as well.

Reply:
Done